libcxx/include/__filesystem/directory_iterator.h

```diff
@@ -73,7 +73,8 @@ class directory_iterator {
   _LIBCPP_HIDE_FROM_ABI ~directory_iterator() = default;
 
   _LIBCPP_HIDE_FROM_ABI const directory_entry& operator*() const {
-    _LIBCPP_ASSERT_UNCATEGORIZED(__imp_, "The end iterator cannot be dereferenced");
+    // Note: this check duplicates a check in `__dereference()`.
```

Comment:
Per our discussion just now, these are the different ways we can think of handling the situation of "redundant" checks:

```c++
// Option #1: leave it as-is
void f(std::optional<T> foo) {
    _LIBCPP_ASSERT_VALID_ELEMENT_ACCESS(foo.has_value(), "oops");
    use(*foo);
}

// Option #2: remove it, it's implicitly checked, we "know" it
void f(std::optional<T> foo) {
    use(*foo);
}

// Option #3: Use a comment
void f(std::optional<T> foo) {
    // implicit precondition: foo.has_value(), already checked in operator* below
    use(*foo);
}

// Option #4: Macro orthogonal to the assertion category
void f(std::optional<T> foo) {
    _LIBCPP_REDUNDANT_ASSERTION(_LIBCPP_ASSERT_VALID_ELEMENT_ACCESS(foo.has_value(), "oops"));
    use(*foo);
}

// Option #5: Add a new category
void f(std::optional<T> foo) {
    // REDUNDANT|EARLY|EXTRA|...
    _LIBCPP_ASSERT_EARLY(foo.has_value(), "oops");
    use(*foo);
}
```

Reply:
Discussion just now, we discussed this by elimination:
- We think we should not do `#1` because the fast mode should be fast, so we should remove duplicate checks
- `#2` is just going to make us scratch our head in the future, it seems that `#3` is strictly better in that case
- `#3` is OK, but we lose the property of having early diagnostics with more meaningful error messages in the case where the extensive mode is enabled. So it is worse than `#4` and `#5`.
- `#4` has the benefit over `#5` that we can retain the real category of the assertion, so it's kinda cleaner in that way. But either would work.

It seems we have a preference for `#4` over `#5`, although not an extremely strong one.